SW-3696 Services & redux for observations in the FE

karthikbtf opened the pull request:
- this is off Steve's [BE PR](https://github.com/terraware/terraware-server/pull/1183), the FE PR will not build in staging
- generated schema and added types/models
- updated existing code to match some breaking model changes on planting sites
- added a service layer that makes one call for now (fetch full hierarhical observations results data)
- tested locally on an org that has observations (generated using Steve's BE script)
- see [gist](https://gist.github.com/karthikbtf/e6a282831ab94414fe7af14eed2a1d81) of what end results looks like in my org
- we may need more services/redux on top of the one API in the service, those will not be covered in this PR.. the current single API covers all the observations page-view requirements, plants dashboard and observation settings related API will need to be separate PRs when the time comes
- created Redux layers
- created initial view of observations using placeholder content
- have observations page fetch real data to decide when to show empty content vs data view


<img width="889" alt="Terraware App 2023-06-02 14-39-03" src="https://github.com/terraware/terraware-web/assets/1865174/9da2cc08-f77e-4703-8d05-27d698c5f531">

charlie-tf reviewed (commented):
Looks good so far

karthikbtf commented:
> Looks good so far

Thanks, will be working on the redux layer next. Going to start simple here and won't add any utilities etc. We can evolve them as we see fit.

karthikbtf commented:
@charlie-tf you could if you wish,, branch off this and create the map component for observations.
You would have to use steve's branch for the BE locally.

karthikbtf commented on src/components/Observations/ObservationsView.tsx:
```diff
@@ -0,0 +1,38 @@
+  selectedPlantingSiteId: number;
+};
+
+export default function ObservationsView({ selectedPlantingSiteId }: ObservationsViewProps): JSX.Element {
```
I will rename this to ObservationsDataView - that's the representation here

karthikbtf commented on src/components/Observations/index.tsx:
```diff
@@ -41,10 +67,12 @@ export default function PlantsDashboard(): JSX.Element {
-        <div>placeholder for selected planting site {selectedPlantingSite?.id}</div>
+      {observationsResults === undefined && observationsResultsError === undefined ? (
+        <CircularProgress sx={{ margin: 'auto' }} />
+      ) : observationsResults?.length && selectedPlantingSite ? (
```
i have to update this,, show empty state if a specific selected planting site has no observations.

charlie-tf commented on src/redux/features/observations/observationsSelectors.ts:
```diff
@@ -0,0 +1,19 @@
+import { createSelector } from '@reduxjs/toolkit';
+
+export const selectObservationsResults = (state: RootState) => state.observationsResults?.observations;
+export const selectObservationsResultsError = (state: RootState) => state.observationsResults?.error;
```
`state.observationResults` comes from `name: 'observationResults'` in the slice file?

charlie-tf commented on src/redux/features/observations/observationsSelectors.ts:
```diff
@@ -0,0 +1,19 @@
+);
+
+export const selectPlantingSiteObservationsResults = createSelector(
+  [selectCompletedObservationsResults, (state, plantingSiteId) => plantingSiteId],
```
What does this list do?

karthikbtf replied:
The observations page has a planting site selector. This list returns observations for the selected planting site - including 'All'.

karthikbtf replied:
One thing about using selectors is that we use selectors to access data. That includes narrowing down of data.

charlie-tf replied:
Got it, Thanks

karthikbtf commented on src/redux/rootReducer.ts:
```diff
@@ -1,9 +1,11 @@
 // assembly of app reducers
 export const reducers = {
   appVersion: appVersionReducer,
+  observationsResults: observationsResultsReducer,
```
@charlie-tf the `state.observationResults` comes from this key

charlie-tf replied:
Ohh, gotcha 👍

karthikbtf commented:
> This is looking good. Just a couple minor questions about the setting up of the selectors

Thanks. I will rename the slice to avoid confusion, and update the completedTime to `yyyy-mm-dd` format with planting site time zone accounted for (so downstream code needn't worry about it).

karthikbtf commented:
> > This is looking good. Just a couple minor questions about the setting up of the selectors
> 
> Thanks. I will rename the slice to avoid confusion, and update the completedTime to `yyyy-mm-dd` format with planting site time zone accounted for (so downstream code needn't worry about it).

Updated PR to map all completed times to planting site time zone. Also renamed the slice.
If the BE PR has any updates that require changes here, I will follow up (won't merge until then ofc).

<img width="559" alt="Terraware App 2023-06-05 11-17-15" src="https://github.com/terraware/terraware-web/assets/1865174/444ba14a-24fc-4833-a286-53a1df015416">

charlie-tf reviewed (approved):
Changes look good 👍

karthikbtf commented on src/redux/features/observations/observationsThunks.ts:
```diff
@@ -1,14 +1,63 @@
+          observations: observations.map((observation) =>
+            applyTimeZone(
+              observation,
+              plantingSiteTimeZones.find((site) => observation.plantingSiteId === site.id)?.timeZone
```
The timezone?

charlie-tf replied:
Yes, that's what I meant, sorry. Will there be a case in which it cannot find a timezone for the plantingSite. I was wondering how the applyTimeZone method handled the case where it was undefined. Digging into it a little, I see that it defaults to `Etc/UTC`. No changes or anything needed, I was just curious
